<?xml version="1.0"?>
<robot name="umi-rtx">
  <!-- Base Link -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="1.252 0.132 0.091"/>
      </geometry>
      <origin rpy="0 -1.57 1.57" xyz="0 -0.0455 0"/>
      <material name="blue">
        <color rgba="0 0 .8 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="1.252 0.132 0.091"/>
      </geometry>
      <origin rpy="0 -1.57 1.57" xyz="0 -0.0455 0"/>
    </collision>
  </link>

  <!-- Shoulder Up-Down Joint -->
  <joint name="shoulder_updown" type="prismatic">
    <parent link="base_link"/>
    <child link="shoulder_link"/>
    <origin xyz="0 0.0445 -0.3" rpy="0 0 1.57"/>
    <axis xyz="0 0 1"/>
    <limit lower="0.033" upper="0.948" effort="1" velocity="1"/>
  </joint>

  <!-- Shoulder Link -->
  <link name="shoulder_link">
    <visual>
      <geometry>
        <box size="0.278 0.132 0.091"/>
      </geometry>
      <origin rpy="0 -1.57 0" xyz="0 0 0"/>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.278 0.132 0.091"/>
      </geometry>
      <origin rpy="0 -1.57 0" xyz="0 0 0"/>
    </collision>
  </link>

  <!-- Shoulder Joint -->
  <joint name="shoulder_joint" type="revolute">
    <parent link="shoulder_link"/>
    <child link="shoulder_to_elbow"/>
    <origin xyz="-0.01 0 -0.09"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.57" upper="1.57" effort="1" velocity="1"/>
  </joint>

  <!-- Shoulder to Elbow Link -->
  <link name="shoulder_to_elbow">
    <visual>
      <geometry>
        <box size="0.252 0.10 0.10"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.126 0 0"/>
      <material name="white">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.252 0.10 0.10"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.126 0 0"/>
    </collision>
  </link>

  <!-- Elbow Joint -->
  <joint name="elbow" type="revolute">
    <parent link="shoulder_to_elbow"/>
    <child link="elbow_to_wrist"/>
    <origin xyz="0.252 0 -0.07"/>
    <axis xyz="0 0 1"/>
    <limit lower="-3.14" upper="2.64" effort="1" velocity="1"/>
  </joint>

  <!-- Elbow to Wrist Link -->
  <link name="elbow_to_wrist">
    <visual>
      <geometry>
        <box size="0.252 0.10 0.07"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.126 0 0"/>
      <material name="white">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.252 0.10 0.07"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.126 0 0"/>
    </collision>
  </link>

  <!-- Wrist Joint -->
  <joint name="wrist" type="revolute">
    <parent link="elbow_to_wrist"/>
    <child link="wrist_to_wrist_gripper_connection"/>
    <origin xyz="0.252 0 -0.05"/>
    <axis xyz="0 0 1"/>
    <limit lower="-1.92" upper="1.92" effort="1" velocity="1"/>
  </joint>

  <!-- Wrist to Wrist Gripper Connection Link -->
  <link name="wrist_to_wrist_gripper_connection">
    <visual>
      <geometry>
        <box size="0.075 0.10 0.05"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <material name="white">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.075 0.10 0.05"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
  </link>

  <!-- Wrist Gripper Connection Roll Joint -->
  <joint name="wrist_gripper_connection_roll" type="revolute">
    <parent link="wrist_to_wrist_gripper_connection"/>
    <child link="virtual_link"/>
    <origin xyz="0 0 0"/>
    <axis xyz="-1 0 0"/>
    <limit lower="-3.16" upper="2.3" effort="1" velocity="1"/>
  </joint>

  <!-- Virtual Link -->
  <link name="virtual_link"/>

  <!-- Wrist Gripper Connection Pitch Joint -->
  <joint name="wrist_gripper_connection_pitch" type="revolute">
    <parent link="virtual_link"/>
    <child link="wrist_gripper_connection_to_gripper"/>
    <origin xyz="0 0 0"/>
    <axis xyz="0 -1 0"/>
    <limit lower="-1.71" upper="0.07" effort="1" velocity="1"/>
  </joint>

  <!-- Wrist Gripper Connection to Gripper Link -->
  <link name="wrist_gripper_connection_to_gripper">
    <visual>
      <geometry>
        <box size="0.10 0.10 0.05"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.05 0 0"/>
      <material name="white">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.10 0.10 0.05"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.05 0 0"/>
    </collision>
  </link>

  <!-- Fixed Joint -->
  <joint name="wrist_gripper_connection" type="fixed">
    <parent link="wrist_gripper_connection_to_gripper"/>
    <child link="gripper_base"/>
    <origin xyz="0 0 0"/>
  </joint>

  <!-- Gripper Base Link -->
  <link name="gripper_base">
    <visual>
      <geometry>
        <box size="0.077 0.15 0.078"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.10 -0.02 0.012"/>
      <material name="white">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.077 0.15 0.078"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.10 -0.02 0.012"/>
    </collision>
  </link>

  <!-- Left Finger Link -->
  <link name="left_finger">
    <visual>
      <geometry>
        <box size="0.07 0.015 0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.06 -0.04 0"/>
      <material name="white">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.07 0.015 0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.06 -0.04 0"/>
    </collision>
  </link>

  <!-- Right Finger Link -->
  <link name="right_finger">
    <visual>
      <geometry>
        <box size="0.07 0.015 0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.06 0.04 0"/>
      <material name="white">
        <color rgba="0.5 0.5 0.5 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="0.07 0.015 0.025"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.06 0.04 0"/>
    </collision>
  </link>

  <!-- Gripper Left Joint -->
  <joint name="gripper_left" type="prismatic">
    <parent link="gripper_base"/>
    <child link="left_finger"/>
    <origin rpy="0 0 0" xyz="0.12 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="0" upper="0.05" effort="1" velocity="1"/>
  </joint>

  <!-- Gripper Right Joint -->
  <joint name="gripper_right" type="prismatic">
    <parent link="gripper_base"/>
    <child link="right_finger"/>
    <origin rpy="0 0 0" xyz="0.12 0 0"/>
    <mimic joint="gripper_left" multiplier="-1"/>
    <axis xyz="0 1 0"/>
    <limit lower="-0.05" upper="0" effort="1" velocity="1"/>
  </joint>
</robot>

